pkg/loadbalancer/stream/loadbalancer.go

```diff
@@ -340,31 +341,39 @@ func (lb *LoadBalancer) setTargets(targets []*nspAPI.Target) error {
 		return nil
 	}
 	var errFinal error
-	toRemoveTargetsMap := make(map[int]struct{})
-	for identifier := range lb.targets {
-		toRemoveTargetsMap[identifier] = struct{}{}
-	}
-	lb.logger.V(2).Info("setTargets", "targets", targets)
-	for _, target := range targets { // targets to add
+	newTargetsMap := make(map[int]types.Target)
+	for _, target := range targets {
 		t, err := NewTarget(target, lb.netUtils)
 		if err != nil {
 			continue
 		}
-		if lb.targetExists(t.GetIdentifier()) {
-			delete(toRemoveTargetsMap, t.GetIdentifier())
-		} else {
-			err = lb.AddTarget(t) // todo: pending targets?
+		newTargetsMap[t.GetIdentifier()] = t
+	}
+	for identifier, target := range lb.targets { // targets to remove
+		newTarget, exists := newTargetsMap[identifier]
+		if !exists {
+			err := lb.RemoveTarget(identifier)
 			if err != nil {
 				errFinal = fmt.Errorf("%w; %v", errFinal, err)
 			}
+			continue
 		}
-	}
-	for identifier := range toRemoveTargetsMap { // targets to remove
+		if slice.ContainsSameStrings(target.GetIps(), newTarget.GetIps()) { // have the same IPs?
```

Comment:
This is a set-operation. The right way is to add a set package, not a slice package. Here are generic sets as implemented in K8s

https://github.com/kubernetes/apimachinery/blob/v0.26.1/pkg/util/sets/set.go#L24

target.GetIps() should really return a set (set[string]). This would make the code more maintainable. This may be a large update, but to implement (steal) a set package may be a good first step.

Reply:
Godoc https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#Set

Reply:
For ref https://github.com/golang/go/issues/57433